Make `botocore` an optional dependency

gadomski opened the pull request:
~Required because `odc.stac.configure_s3_access` is part of the public API, which imports from `_aws` which requires `botocore`.~

Adds `botocore` as an optional dependency. Includes:

- Nicer import error message when `botocore` is not installed
- Skip aws tests when botocore is not installed
- CI with/without botocore
- README instructions on using `extra_requires` with pip

gadomski committed:
feat: add botocore to as an extra_requires

Also includes CI test w/o botocore and a friendlier import error message.

Kirill888 commented:
@gadomski I don't see `botocore` as an essential dependency, it's use is (or at least should be) optional, boto is a giant dependency and if don't care for it you should be able to skip it. `rasterio` for example uses boto3 when installed but doesn't declare it as a compulsory dependency. Adding it behind a feature flag is an option though.

gadomski commented:
> Adding it behind a feature flag is an option though.

This makes sense to me, maybe with a graceful catch and friendly message on an `ImportError`>

Kirill888 reviewed (changes requested):
should be behind feature flag, imports from `_aws` only happen on first use inside `configure_s3_access` function, `odc-stac` should work fine without botocore installed, this was intentional choice.

Would not hurt adding import failure guards with reasonable warning though.

Kirill888 commented:
@gadomski do you want to add `s3` feature to setup before releasing 0.3.1 or should we just release the fix?

gadomski commented:
I think this PR is pretty minor, I think v0.3.1 without this is fine ... I've kicked off for the day so my changes won't come until tomorrow in the US.

Kirill888 commented:
ok, I'll make a release then

gadomski commented:
Also might be worth setting up branch protection, looks like I can merge this even though you requested changes.

Kirill888 commented on tests/test_utils_aws.py:
```diff
@@ -30,6 +39,18 @@ def mock_urlopen(text, code=200):
     return m
 
 
+def requires_botocore(f: Callable) -> Callable:
```
Since all functions here need botocore to run probably easier to just use:

https://docs.pytest.org/en/stable/reference/reference.html#pytest-importorskip

gadomski replied:
Nice, wasn't aware of that function, thanks!

Kirill888 commented:
current test environment used on Actions already contains botocore because it's pulled in by datacube (when installed with conda), we'll need a separate environment for testing minimal install of `odc-stac` and this will most likely require protecting more tests with import guards.

Kirill888 commented on odc/stac/_aws.py:
```diff
@@ -5,16 +5,25 @@
 """
```
Not too keen on those changes. The idea was that `_aws.py` is a private module that contains only things that are meaningful when `botocore` is installed in the environment. The only place that this module is imported, apart for tests, is here:

https://github.com/opendatacube/odc-stac/blob/1c6fd02b6a95f5ebca3da9fb3dfe4df210eae105/odc/stac/_rio.py#L214-L221

https://github.com/opendatacube/odc-stac/blob/1c6fd02b6a95f5ebca3da9fb3dfe4df210eae105/odc/stac/_rio.py#L248-L249

It's the only public interface function that interacts with that file.

So that's where import guards should go. This is how we dealing with this in odc-geo, but there is no equivalent structure in odc-stac

https://github.com/opendatacube/odc-geo/blob/4f9004720d899dff4bf0b208e4fe602790d055d7/odc/geo/_interop.py#L33-L39

gadomski replied:
What is your preferred behavior? Remove `get_aws_settings` from the public interface if `botocore` isn't available?

Kirill888 replied:
it's not part of public interface though, `configure_s3_access` is. 

I think best way to deal with this is to keep `configure_s3_access`  public but raise error message if used without `botocore` installed, but do all that inside `_rio.py` keeping `_aws.py` as it is now, without any guards and meta programming that can interfere with types.

If you think it will make it clearer, feel free to rename every function inside _aws to start with `_`, to me these are "public functions within a private module" so not prefixed with `_`.

Kirill888 replied:
These are public:

https://github.com/opendatacube/odc-stac/blob/1c6fd02b6a95f5ebca3da9fb3dfe4df210eae105/odc/stac/__init__.py#L16-L27

and also these when "datacube" is installed:

https://github.com/opendatacube/odc-stac/blob/1c6fd02b6a95f5ebca3da9fb3dfe4df210eae105/odc/stac/eo3/__init__.py#L9-L10

and finally "bench" tools, which need extra libs too

https://github.com/opendatacube/odc-stac/blob/1c6fd02b6a95f5ebca3da9fb3dfe4df210eae105/odc/stac/bench/__init__.py#L14-L22

gadomski commented:
I've updated this PR to be much simpler. I got a little turned around because I didn't realize `_aws` was only being imported in that one `_rio` function -- I was just looking at the unit tests. I've updated the PR description with the current approach.

Kirill888 reviewed (approved):
thanks, much cleaner now
